### ISSUE TITLE
Failure to connect due to using the wrong Port Number

### ISSUE THREAD

**Issue by OscarPhilips** [Hugovdberg/PIconnect#535]
Bug report
==========

Description
-----------

I am getting a TCP/IP Port connection failure … see below:

When I build my Python script (Anaconda 3.7.7) in Spyder, using PIConnect, everything works like a charm down though this point:

### IN SPYDER
```
import` sys
import re
import PIconnect as PI
from PIconnect.PIConsts import SummaryType, TimestampCalculation

with PI.PIServer(server='mypi') as server:
    print(server.server_name)
```
This builds a connection to my PI server (mypi) over port 5450 as I would expect, and all is well with the world.
 
However, when I try to run Python from the command line, using either python.exe or ipython.exe (again Anaconda 3.3.7) the connection fails as it tries to connect on port 5457.

```
C:\Users\14541\OneDrive - FirstEnergy Corp\My Python\OSI PI>python.exe test.py
System.Net.Sockets.SocketException: No connection could be made because the target machine actively refused it 10.10.16.99:5457
   at System.Net.Sockets.Socket.DoConnect(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.Sockets.Socket.Connect(EndPoint remoteEP)
   at System.ServiceModel.Channels.SocketConnectionInitiator.Connect(Uri uri, TimeSpan timeout)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test.py", line 3, in <module>
    import PIconnect as PI
  File "D:\Anaconda3\lib\site-packages\PIconnect\__init__.py", line 28, in <module>
    from PIconnect.PIAF import PIAFDatabase
  File "D:\Anaconda3\lib\site-packages\PIconnect\PIAF.py", line 61, in <module>
    class PIAFDatabase(object):
  File "D:\Anaconda3\lib\site-packages\PIconnect\PIAF.py", line 71, in PIAFDatabase
    for s in AF.PISystems()
  File "D:\Anaconda3\lib\site-packages\PIconnect\PIAF.py", line 71, in <dictcomp>
    for s in AF.PISystems()
  File "D:\Anaconda3\lib\site-packages\PIconnect\PIAF.py", line 70, in <dictcomp>
    s.Name: {"server": s, "databases": {d.Name: d for d in s.Databases}}
System.ServiceModel.EndpointNotFoundException: Could not connect to net.tcp://mypi:5457/AFServer/Service. The connection attempt lasted for a time span of 00:00:01.1774284. TCP error code 10061: No connection could be made because the target machine actively refused it 10.10.16.99:5457.

Server stack trace:
   at System.ServiceModel.Channels.SocketConnectionInitiator.Connect(Uri uri, TimeSpan timeout)
   at System.ServiceModel.Channels.BufferedConnectionInitiator.Connect(Uri uri, TimeSpan timeout)
   at System.ServiceModel.Channels.ConnectionPoolHelper.EstablishConnection(TimeSpan timeout)
   at System.ServiceModel.Channels.ClientFramingDuplexSessionChannel.OnOpen(TimeSpan timeout)
   at System.ServiceModel.Channels.CommunicationObject.Open(TimeSpan timeout)
   at System.ServiceModel.Channels.ServiceChannel.OnOpen(TimeSpan timeout)
   at System.ServiceModel.Channels.CommunicationObject.Open(TimeSpan timeout)

Exception rethrown at [0]:
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at System.ServiceModel.ICommunicationObject.Open(TimeSpan timeout)
   at System.ServiceModel.ClientBase`1.System.ServiceModel.ICommunicationObject.Open(TimeSpan timeout)
   at OSIsoft.AF.Support.AFProxy.Reconnect(Boolean autoPrompt, AFConnectionProtocol protocol, String host, Int32 port, String accountName, TimeSpan timeout)

C:\Users\14541\OneDrive - FirstEnergy Corp\My Python\OSI PI>

```
Expected behavior
-----------------

I would expcect a connection to my PI server, mypi:5450

Possible solutions
------------------

Screenshots
-----------

All is set up in the OSI PI SDK Utility correctly, including Server Name and Port Number

System
------

-   OS: Windows 10 (64 bit)
-   Python version: Anaconda Python 3.7.7
-   PIconnect version: 0.8.0

Additional context
------------------

**Comment by Hugovdberg** [comment]
This is an annoying error indeed, which I have had happen a few times as well with a hanging server. This also occurs if one the configured servers is unreachable. There is a fix in the making to at least be able to import the package. However, I don't know which ports are in use exactly, the ports are handled entirely by the OSISoft SDK, so I don't know if that fix will help you.

**Comment by FCCMac** [comment]
Any more word on this? I'm running into the same issue now. I had been accessing our servers with no issue through a Jupyter Notebook, but attempting to import PIconnect in a script gives the same error @OscarPhilips mentioned:

    System.ServiceModel.EndpointNotFoundException: Could not connect to net.tcp://<pi-server>:5457/AFServer/Service.
    The connection attempt lasted for a time span of 00:00:02.1100165.
    TCP error code 10061: No connection could be made because the target machine actively refused it <server_ip>:5457.

**Comment by fouadsel88** [comment]
Hi I am getting similar error with: **System.Net.Sockets.SocketException: No such host is known** is there any fix for that?
I tried Notebook too and  didn't work.

**Comment by somuchung** [comment]
Has anyone found a work around this?

**Comment by Hugovdberg** [comment]
Sorry guys, I haven't come around to fixing this lately. I mentioned I was working on a fix, but I can't find where I was working on that. I guess it's time to do some work on this package, there's some dependencies that need updating as well.

**Comment by OscarPhilips** [comment]
I originally posted the issue of wrong port numbers being used back in July 2020.  I just returned to my project in late January 2021, updated my Anaconda environment and now it works ... at least for now.

Same code as before, no edits or changes, so I cannot put my finger on what might have changed.

**Comment by OscarPhilips** [comment]
My current work around it to 

1. try to connect to one PI server, on first run, it usually fails, 
2. try to connect to another PI server, it also fails
3. try to connect to the first PI server again, it usually works.

Now that is strange.  I can beat on one server all night and "no good," but toggle servers back and forth and I am in.  I am not sure what you do if you only have access to one PI server.

**Comment by FCCMac** [comment]
@OscarPhilips Sure enough, this works for me. I started up a fresh notebook in VS Code and was getting the exception on `import PIconnect as PI`, switched to our second HA server, still get the exception at import, then switched back to our primary and the import statement now works. As I work today, I'll keep you posted about how long the fix lasts to see if that might give us some insights as to why it's happening.

**Comment by Hugovdberg** [comment]
This is really weird, I cannot reproduce it and have no idea how to get a
clue why this happens to you guys. PIconnect simply connects using the
default configuration that the SDK is configured with. Does the same thing
happen when you use another application that uses the SDK to communicate
(which I believe is every other PI client software such as the Excel
datelink or process book)? I just ask the SDK which servers it knows and
connect to the default one if you don't specify otherwise.

Op di 2 feb. 2021 19:02 schreef Chris M. <notifications@github.com>:

**Comment by papabasile** [comment]
I've had the same issue. There are some servers on my SDK that can't be accessed (need to vpn) and I've set the default server to the one I need to access and have currently access to. However PIconnect import still tries to connect to another server and fails. It seems repeated import calls within a notebook rotate to all the available servers and eventually import works.

**Comment by Hugovdberg** [comment]
@OscarPhilips @FCCMac @fouadsel88 @somuchung @papabasile 
Could you guys please try to install the latest version (0.9.1) using pip to see if these errors are fixed now? If so, we can finally close this issue.

**Comment by Hugovdberg** [comment]
Closing for now, please reopen or create a new issue if it isn't fixed after installing v0.9.1.

**Comment by OscarPhilips** [comment]
Sorry, it took me a while to get to testing, but all the problems connecting I was experiencing are resolved.   Thank you for your efforts.
